### PR TITLE
[GCCH] Company Communicator in GCCH

### PR DESCRIPTION
**Pull request by juna-gupta**
**Problem**
Company Communicator code in current form will not work against Teams GCCH / DOD environment. We want to update the solution to make it work in these environments. This PR attempts to make minimal changes required to do so. 

**Solution:**
This is the first of 2 PRs I plan to publish for Company Communicator to work in GCCH. 

This change covers the following:
1. Updated Bot Adapter to extend from Cloud Adapter, and related libraries.
2. Updated Graph, AAD login, Identity library config against Gov environment.
3. Teams license ID in GCCH.
4. Added app configuration to use common code base for different environments.

**Validations:**
End user scenarios - Send message to all users, members of a team (proactive installation and messaging).
App Id / App secret flow. (did not test KeyVault, App certificates)

**Next steps**
Validate pending flows.

### CODE REVIEW

File: Source/CompanyCommunicator.Common/Adapter/CCBotAdapterBase.cs [hunk 1]
```diff
@@ -8,16 +8,26 @@ namespace Microsoft.Teams.Apps.CompanyCommunicator.Common.Adapter
...
     using Microsoft.Bot.Builder.Integration.AspNet.Core;
     using Microsoft.Bot.Connector.Authentication;
     using Microsoft.Bot.Schema;
+    using Microsoft.Extensions.Logging;
 
     /// <summary>
     /// Bot Framework Http Adapter interface.
     /// </summary>
-    public interface ICCBotFrameworkHttpAdapter : IAdapterIntegration, IBotFrameworkHttpAdapter
+    public abstract class CCBotAdapterBase : CloudAdapter
```
Comment (aosolis) [inline Source/CompanyCommunicator.Common/Adapter/CCBotAdapterBase.cs:19]: What's the purpose of interposing this abstract class in the hierarchy?

File: Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        /// <summary>
+        /// Configuration factory returns relevant configuration for a given environment.
+        /// </summary>
+        /// <param name="env">Teams environment.</param>
+        /// <returns>App configurstion.</returns>
+        public IAppConfiguration GetAppConfiguration(string env)
+        {
+            switch (env)
+            {
+                case "GCCH":
```
Comment (aosolis) [inline Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs]: Should this be in an enum to ensure that this covers all environments?

File: Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                    return new GCCHConfiguration(this.tenantId);
+                case "DOD":
+                    return new DODConfiguration(this.tenantId);
+                case "Commercial":
```
Comment (aosolis) [inline Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs]: We should probably have GCC here too, even though it's the same as Commercial.

File: Source/CompanyCommunicator.Common/Configuration/DODConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        public string AzureAd_Instance => "https://login.microsoftonline.us/";
+
+        /// <inheritdoc/>
+        public string AzureAd_ValidIssuers => $"https://login.microsoftonline.us/{this.tenantId}/v2.0,https://sts.windows.net/{this.tenantId}/";
+
+        /// <inheritdoc/>
+        public string AuthorityUri => $"https://login.microsoftonline.us/{this.tenantId}";
+
+        /// <inheritdoc/>
+        public string GraphBaseUrl => "https://graph.microsoft.us/v1.0";
```
Comment (aosolis) [inline Source/CompanyCommunicator.Common/Configuration/DODConfiguration.cs]: Is this correct? The docs say https://dod-graph.microsoft.us (https://learn.microsoft.com/en-us/graph/deployments)

File: Source/Test/CompanyCommunicator.Test/Controllers/AuthenticationMetadataControllerTest.cs [hunk 5]
```diff
@@ -47,7 +47,7 @@ public void CreateInstance_AllParameters_ShouldBeSuccess()
         {
             // Arrange
             this.options.Setup(x => x.Value).Returns(new AuthenticationOptions() { AzureAdTenantId = this.tenantId, AzureAdClientId = this.clientId });
-            Action action = () => new AuthenticationMetadataController(this.options.Object);
+            Action action = () => new AuthenticationMetadataController(this.options.Object, null /*TODO - fix this*/);
```
Comment (aosolis) [inline Source/Test/CompanyCommunicator.Test/Controllers/AuthenticationMetadataControllerTest.cs:50]: what's the resolution for these TODOs

**Review by aosolis — CHANGES_REQUESTED** [review]
See comments

File: Source/CompanyCommunicator.Common/Adapter/CCBotAdapterBase.cs (hunk 1 above)
Comment (juna-gupta) [reply Source/CompanyCommunicator.Common/Adapter/CCBotAdapterBase.cs:19]: In the current state, we had 2 classes that were implementing the interface. Now that we're extending CloudAdapter, switched to an abstract class.

File: Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs (hunk 3 above)
Comment (juna-gupta) [reply Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs]: Yeah, I'll add.

File: Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs (hunk 2 above)
Comment (juna-gupta) [reply Source/CompanyCommunicator.Common/Configuration/ConfigurationFactory.cs]: We could use an enum. Would need an additional change to convert string to enum when reading config.

File: Source/CompanyCommunicator.Common/Configuration/DODConfiguration.cs (hunk 4 above)
Comment (juna-gupta) [reply Source/CompanyCommunicator.Common/Configuration/DODConfiguration.cs]: it should be "dod-graph". Thanks for catching that.

File: Source/Test/CompanyCommunicator.Test/Controllers/AuthenticationMetadataControllerTest.cs (hunk 5 above)
Comment (juna-gupta) [reply Source/Test/CompanyCommunicator.Test/Controllers/AuthenticationMetadataControllerTest.cs:50]: I fixed all the build issues in the test projects. These tests need to be updated. Have asked the template team to help with that effort.

**Review by aosolis — APPROVED** [review]

**Pull request merged**